src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -1380,11 +1380,26 @@ internal protected virtual TypeDefinition MarkType (TypeReference reference, Dep
 				MarkStaticConstructor (type, new DependencyInfo (DependencyKind.TriggersCctorForCalledMethod, reason.Source), type);
 
 			// Check type being used was not removed by the LinkerRemovableAttribute
-			if (_context.Annotations.HasLinkerAttribute<RemoveAttributeInstancesAttribute> (type))
-				_context.LogWarning ($"Custom Attribute {type.GetDisplayName ()} is being referenced in code but the linker was " +
-					$"instructed to remove all instances of this attribute. If the attribute instances are necessary make sure to " +
-					$"either remove the linker attribute XML portion which removes the attribute instances, or to override this use " +
-					$"the linker XML descriptor to keep the attribute type (which in turn keeps all of its instances).", 2045, sourceLocationMember);
+			if (_context.Annotations.HasLinkerAttribute<RemoveAttributeInstancesAttribute> (type)) {
+				bool shouldWarn = true;
+
+				// Don't warn about references from code which is not linked
+				//if (Annotations.GetAction (sourceLocationMember.DeclaringType.Module.Assembly) != AssemblyAction.Link)
+				//	shouldWarn = false;
```

Comment:
I'm not sure we should suppress the warning in this case. The scenario is basically:
* An attribute is marked for removal (let's say attribute `ToBeRemovedAttribute`)
* The attribute instantiated in some places in linked assemblies - and those instantiations will be removed
* There's an assembly which is not linked ("copy") which has a type reference to the attribute type - so for example `typeof(ToBeRemovedAttribute)`.

The question is, should this type reference generate a 2045 warning even if it's from an assembly which is not linked?

My take is: It should - the code is likely to break if it uses the attribute's type to look up attribute instances as those won't be there. The fact that it's in a copy assembly is not different than if it would be a linked assembly, but in code which is reachable.

I intentionally left the code commented out and a TODO - once we agree on the answer I'll update the PR.

Reply:
Will this case also be hit for a "copy" or "copyused" assembly that has an instance of `ToBeRemovedAttribute` (rather than referencing it via `typeof`)? I am wondering how often this is likely to show up in user scenarios.

Reply:
Having an instance doesn't trigger this warning - that's exactly what happened with the warnings in CoreLib you reported - copy assemblies had instances of some of the attributes  - the instance itself doesn't cause warnings even without this change. It does cause warnings within the attribute type - which is what this change fixes.

Reply:
In that case I agree we should emit the warning even for non-linked assemblies.